ArduPlane/quadplane.cpp

```diff
@@ -2267,50 +2288,109 @@ void QuadPlane::vtol_position_controller(void)
+                                                           plane.ahrs.yaw_sensor, false);
+
+        if (distance < stop_distance) {
+            gcs().send_text(MAV_SEVERITY_INFO,"VTOL position1 started v=%.1f d=%.1f",
+                            (double)groundspeed, (double)plane.auto_state.wp_distance);
+            poscontrol.state = QPOS_POSITION1;
+        }
+        break;
+    }
+
     case QPOS_POSITION1: {
         const Vector2f diff_wp = plane.current_loc.get_distance_NE(loc);
         const float distance = diff_wp.length();
         Vector2f groundspeed = ahrs.groundspeed_vector();
         float speed_towards_target = distance>1?(diff_wp.normalized() * groundspeed):0;
-        if (poscontrol.speed_scale <= 0) {
-            // initialise scaling so we start off targeting our
-            // current linear speed towards the target. If this is
-            // less than the wpnav speed then the wpnav speed is used
-            // land_speed_scale is then used to linearly change
-            // velocity as we approach the waypoint, aiming for zero
-            // speed at the waypoint
-            // setup land_speed_scale so at current distance we
-            // maintain speed towards target, and slow down as we
-            // approach
-
-            // max_speed will control how fast we will fly. It will always decrease
-            poscontrol.max_speed = MAX(speed_towards_target, wp_nav->get_default_speed_xy() * 0.01);
-            poscontrol.speed_scale = poscontrol.max_speed / MAX(distance, 1);
+
+        // Initialise position control to target zero velocity
+        // above the waypoint.
+        if (poscontrol.stopping_distance <= 0) {
+            // Calculate stopping distance based on the current groundspeed
+            // and the transition deceleration target
+            poscontrol.stopping_distance = groundspeed.length_squared()/(2*transition_decel);
+            poscontrol.pre_decel_gndspd = groundspeed.length();
+            
+            float throttle = SRV_Channels::get_output_scaled(SRV_Channel::k_throttle);
+            if (!plane.have_reverse_thrust()) {
+                vel_forward.integrator = constrain_int16(throttle, 0, 100);
+            } else {
+               vel_forward.integrator = constrain_int16(throttle, -100, 100);  
+            }
+
+            vel_forward.last_ms = AP_HAL::millis();
         }
 
         // run fixed wing navigation
         plane.nav_controller->update_waypoint(plane.prev_WP_loc, loc);
 
-        /*
-          calculate target velocity, not dropping it below 2m/s
-         */
-        const float final_speed = 2.0f;
-        Vector2f target_speed_xy = diff_wp * poscontrol.speed_scale;
+        Vector2f target_speed_xy = diff_wp.normalized() * poscontrol.pre_decel_gndspd;
         float target_speed = target_speed_xy.length();
+
+        // if vehicle is within deceleration profile, update target speed
+        if (poscontrol.stopping_distance > distance) {
```

Comment:
Why do we stash this value instead of calculating it on demand? I don't see any reason we should be assuming the stopping distance calculation is still correct. If we were approaching from 1 km away the wind conditions could be very different and the distance calculation could be completely wrong, not to mention we may have been going the wrong way when we first calculated it and have had a tail wind, but by the time we reach here it we may have switched to a head wind which would cause us to detransition far to early (the other way around is even worse for transitioning)

Reply:
This was an adjustment made by Luke Dempsey and was for the previous code where we may have started POSITION1 before the stopping distance. I've removed it now, as we don't start POSITION1 until we've reached the stopping distance

Reply:
I think this still needs more work though, to account for the "going the wrong way" case

Reply:
fixed